components/client/src/main/java/com/hotels/styx/client/Transport.java

```diff
@@ -43,37 +42,31 @@ public Transport(Id appId, CharSequence originIdHeaderName) {
     public HttpTransaction send(LiveHttpRequest request, Optional<ConnectionPool> origin, Id originId) {
         Observable<Connection> connection = connection(request, origin);
 
-        AtomicReference<Connection> connectionRef = new AtomicReference<>(null);
-        Observable<LiveHttpResponse> observableResponse = connection.flatMap(tConnection -> {
-            connectionRef.set(tConnection);
-            return tConnection.write(request)
-                    .map(response -> addOriginId(originId, response));
-        });
-
         return new HttpTransaction() {
             @Override
             public Observable<LiveHttpResponse> response() {
-                return ResponseEventListener.from(observableResponse)
-                        .whenCancelled(() -> closeIfConnected(origin, connectionRef))
-                        .whenResponseError(cause -> closeIfConnected(origin, connectionRef))
-                        .whenContentError(cause -> closeIfConnected(origin, connectionRef))
-                        .whenCompleted(() -> returnIfConnected(origin, connectionRef))
-                        .apply();
+                return connection.flatMap(connection -> {
+                    Observable<LiveHttpResponse> responseObservable = connection.write(request)
+                            .map(response -> addOriginId(originId, response));
+
+                    return ResponseEventListener.from(responseObservable)
+                            .whenCancelled(() -> closeIfConnected(origin, connection))
+                            .whenResponseError(cause -> closeIfConnected(origin, connection))
+                            .whenContentError(cause -> closeIfConnected(origin, connection))
+                            .whenCompleted(() -> returnIfConnected(origin, connection))
+                            .apply();
+                });
```

Comment:
Hi @taer 

The observable should be constructed like this:

```            @Override
            public Observable<LiveHttpResponse> response() {
                return connection.flatMap(connection -> {
                    Observable<LiveHttpResponse> responseObservable = connection.write(request)
                            .map(response -> addOriginId(originId, response));

                    return ResponseEventListener.from(responseObservable)
                            .whenCancelled(() -> closeIfConnected(origin, connection))
                            .whenResponseError(cause -> closeIfConnected(origin, connection))
                            .whenContentError(cause -> closeIfConnected(origin, connection))
                            .whenCompleted(() -> returnIfConnected(origin, connection))
                            .apply();
                });
            }
```

The `ResponseEventListener.from` takes a response observable and decorates it with event listeners triggering  at relevant points of the response lifecycle. It takes into account both response observable and its associated body byte stream. The `apply` method returns a new `Observable<LiveHttpResponse>` that should be returned for the downstream observers.

Also, I have pushed this to your branch. Please pull the lates changes to have a look!